### ISSUE TITLE
Testing with 100 orchestrators and 1000 activities each, getting this exception: Part10 !!! Assertion Failed: bad fragment sequence

### ISSUE THREAD

**Issue by ericleigh007** [microsoft/durabletask-netherite#231]
i'm doing some load testing comparing Netherite with the default back end and I must say this backend is so much better in terms of latencies and throughput.

Great job.  We are very much looking forward to amping up our Durable Functions performance now that this is GA.

however, I'm getting an error every once in a while when doing some maximum test, in this case 200 orchestrators that each launch off 1000 activities.

The error looks like this
````
2023-03-14T00:32:11Z   [Error]   Part10 !!! Assertion Failed: bad fragment sequence: position in    at System.Environment.get_StackTrace()
   at DurableTask.Netherite.Partition.Assert(Boolean condition, String message) in /_/src/DurableTask.Netherite/OrchestrationService/Partition.cs:line 164
   at DurableTask.Netherite.FragmentationAndReassembly.Reassemble[TEvent](IEnumerable`1 earlierFragments, IEventFragment lastFragment, Partition partition) in /_/src/DurableTask.Netherite/Util/FragmentationAndReassembly.cs:line 86
   at DurableTask.Netherite.ReassemblyState.Process(PartitionEventFragment evt, EffectTracker effects) in /_/src/DurableTask.Netherite/PartitionState/ReassemblyState.cs:line 61
   at DurableTask.Netherite.EffectTracker.ProcessEffectOn(TrackedObject trackedObject) in /_/src/DurableTask.Netherite/Abstractions/PartitionState/EffectTracker.cs:line 80
   at DurableTask.Netherite.Faster.FasterKV.ProcessEffectOnTrackedObject(Key k, EffectTracker tracker) in /_/src/DurableTask.Netherite/StorageLayer/Faster/FasterKV.cs:line 785
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at DurableTask.Netherite.Faster.FasterKV.ProcessEffectOnTrackedObject(Key k, EffectTracker tracker)
   at DurableTask.Netherite.Faster.TrackedObjectStoreEffectTracker.ApplyToStore(TrackedObjectKey key, EffectTracker tracker) in /_/src/DurableTask.Netherite/StorageLayer/Faster/TrackedObjectStoreEffectTracker.cs:line 26
   at DurableTask.Netherite.EffectTracker.<ProcessUpdate>g__ProcessRecursively|27_0() in /_/src/DurableTask.Netherite/Abstractions/PartitionState/EffectTracker.cs:line 128
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at DurableTask.Netherite.EffectTracker.<ProcessUpdate>g__ProcessRecursively|27_0()
   at DurableTask.Netherite.EffectTracker.ProcessUpdate(PartitionUpdateEvent updateEvent) in /_/src/DurableTask.Netherite/Abstractions/PartitionState/EffectTracker.cs:line 112
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at DurableTask.Netherite.EffectTracker.ProcessUpdate(PartitionUpdateEvent updateEvent)
````
To give a little more context, the orchestrators are only passing a few bytes of data back and forth.  I'm not moving actual data through the system, just launching orchestrators in parallel and they are lauching activities in parallel

my host.json settings are
```json
{
  "extensions": {
    "durableTask": {
      "storageProvider": {
        "type": "Netherite"
      },
      "maxConcurrentActivityFunctions": 50,
      "maxConcurrentOrchestratorFunctions": 100
    }
  },
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": true,
        "excludedTypes": "Request"
      }
    }
  },
  "version": "2.0"
}
```

My code is a complete non-proprietary test case, but it requires a bit of Azure umph as one might expect.  My cosmos db is currently scaled to 50kRU autoscale.

if the code is needed, I'll replicate it into my personal account from work account.

**Comment by davidmrdavid** [comment]
Hi @ericleigh007, thanks for reaching out.
I'd be curious to learn more about the reproducer you're using to get this exception. Is there a minimal repro you could describe and/or put together for us to investigate with? Additionally, I'm curious if this exception is causing your orchestrators to fail, or if it's simply a "benign error" in your logs that doesn't actually prevent your workflows to complete. Please let me know, thanks!

**Comment by sebastianburckhardt** [comment]
We definitely want to get to the bottom of this sooner rather than later - the error is not benign, it indicates something is going wrong at a fundamental level.

The first step is for us to be able to repro. I will start by running the type of scenario you are describing (200 orchestrators that each launch off 1000 activities). Were you using one of the hosted Azure Functions plans (e.g. consumption or EP1, EP2, EP3) when you encountered this? Do you know how many worker nodes were running at the time?

**Comment by sebastianburckhardt** [comment]
Looks like I got a repro. Will continue investigating tomorrow.

**Comment by ericleigh007** [comment]
Hi, so far I can confirm that I'm on premium EP3 plan of durable functions.  I selected the standard event hub if that makes any difference.
I can publish all my code in a public repo If you'd like.
I don't know how many worker nodes were running.

I like the Netherite backend for the level of efficieincy especially as compared to the default.  In my scenario I get terrible latencies of minutes from the default backend whilst Netherite gives me just seconds.  Bravo!   Being able to get that kind of efficiency means i don't have to keep scaling just to support the "overhead" of the orchestrators and activities.

I've got another error here, maybe you're getting that as well.  It looks like this (if you would like to have this filed separately, let me know and I'll do it)
```
2023-03-14T23:50:14Z   [Verbose]   Part00 LogWorker completed batch: batchSize=1 elapsedMilliseconds=18.66 nextBatch=
2023-03-14T23:50:14Z   [Warning]   EventHubsProcessor partitions/0 failed to checkpoint receive position: Microsoft.Azure.EventHubs.Processor.LeaseLostException: The lease ID specified did not match the lease ID for the blob.
 ---> Microsoft.Azure.Storage.StorageException: The lease ID specified did not match the lease ID for the blob.
   at Microsoft.Azure.Storage.Core.Executor.Executor.ExecuteAsync[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Microsoft.Azure.Storage.Blob.CloudBlockBlob.UploadFromStreamAsyncHelper(Stream source, Nullable`1 length, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext, AggregatingProgressIncrementer progressIncrementer, CancellationToken cancellationToken)
   at Microsoft.Azure.Storage.Blob.CloudBlockBlob.UploadFromByteArrayAsync(Byte[] buffer, Int32 index, Int32 count, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext, IProgress`1 progressHandler, CancellationToken cancellationToken)
   at Microsoft.Azure.Storage.Blob.CloudBlockBlob.UploadTextAsync(String content, Encoding encoding, AccessCondition accessCondition, BlobRequestOptions options, OperationContext operationContext, IProgress`1 progressHandler, CancellationToken cancellationToken)
   at Microsoft.Azure.EventHubs.Processor.AzureStorageCheckpointLeaseManager.UpdateLeaseCoreAsync(AzureBlobLease lease)
Request Information
RequestID:292637ee-301e-0086-18cf-561bad000000
RequestDate:Tue, 14 Mar 2023 23:50:14 GMT
StatusMessage:The lease ID specified did not match the lease ID for the blob.
ErrorCode:LeaseIdMismatchWithBlobOperation
ErrorMessage:The lease ID specified did not match the lease ID for the blob.
RequestId:292637ee-301e-0086-18cf-561bad000000
Time:2023-03-14T23:50:14.2251702Z

   --- End of inner exception stack trace ---
   at Microsoft.Azure.EventHubs.Processor.AzureStorageCheckpointLeaseManager.UpdateLeaseCoreAsync(AzureBlobLease lease)
   at Microsoft.Azure.EventHubs.Processor.AzureStorageCheckpointLeaseManager.UpdateCheckpointAsync(Lease lease, Checkpoint checkpoint)
   at Microsoft.Azure.EventHubs.Processor.PartitionContext.PersistCheckpointAsync(Checkpoint checkpoint)
   at DurableTask.Netherite.EventHubsTransport.EventHubsProcessor.SaveEventHubsReceiverCheckpoint(PartitionContext context, Int64 byteThreshold) in /_/src/DurableTask.Netherite/TransportLayer/EventHubs/EventHubsProcessor.cs:line 309
```

**Comment by sebastianburckhardt** [comment]
I was able to analyze my local repro and found that this is a bug in the fragmentation/reassembly logic. I am now working on a fix.

**Comment by sebastianburckhardt** [comment]
By the way, the second exception you mentioned above is harmless. That is why it is prefixed as a warning:
```
2023-03-14T23:50:14Z   [Warning]   EventHubsProcessor partitions/0 failed to checkpoint receive position
```

**Comment by sebastianburckhardt** [comment]
Fix was released with 1.3.3.

Let me know if you are still experiencing these issues after upgrading.

**Comment by ericleigh007** [comment]
I thought I had closed this one before, sorry about that.
The problem is gone in 1.3.3.  None of the problems i was seeing previously have re-appeared, so I'll close the issue.
